partiql-types/src/main/kotlin/org/partiql/value/PartiQLValue.kt

```diff
@@ -0,0 +1,425 @@
+
+    abstract override fun withAnnotations(annotations: Annotations): TimeValue
+
+    abstract override fun withoutAnnotations(): TimeValue
+}
+
+public abstract class TimestampValue : ScalarValue<LocalDateTime> {
+
+    override val type: PartiQLValueType = PartiQLValueType.TIMESTAMP
+
+    // TEMPORARY
+    public abstract val timeZone: TimeZone?
+
+    // TEMPORARY
+    public abstract val precision: Int
+
+    abstract override fun copy(annotations: Annotations): TimestampValue
+
+    abstract override fun withAnnotations(annotations: Annotations): TimestampValue
+
+    abstract override fun withoutAnnotations(): TimestampValue
+}
+
+public abstract class IntervalValue : ScalarValue<Long> {
+
+    override val type: PartiQLValueType = PartiQLValueType.INTERVAL
+
+    abstract override fun copy(annotations: Annotations): IntervalValue
+
+    abstract override fun withAnnotations(annotations: Annotations): IntervalValue
+
+    abstract override fun withoutAnnotations(): IntervalValue
+}
+
+public abstract class BagValue<T : PartiQLValue> : CollectionValue<T> {
+
+    override val type: PartiQLValueType = PartiQLValueType.BAG
+
+    abstract override fun copy(annotations: Annotations): BagValue<T>
+
+    abstract override fun withAnnotations(annotations: Annotations): BagValue<T>
+
+    abstract override fun withoutAnnotations(): BagValue<T>
+}
+
+public abstract class ListValue<T : PartiQLValue> : CollectionValue<T> {
+
+    override val type: PartiQLValueType = PartiQLValueType.LIST
+
+    abstract override fun copy(annotations: Annotations): ListValue<T>
+
+    abstract override fun withAnnotations(annotations: Annotations): ListValue<T>
+
+    abstract override fun withoutAnnotations(): ListValue<T>
+}
+
+public abstract class SexpValue<T : PartiQLValue> : CollectionValue<T> {
+
+    override val type: PartiQLValueType = PartiQLValueType.SEXP
+
+    abstract override fun copy(annotations: Annotations): SexpValue<T>
+
+    abstract override fun withAnnotations(annotations: Annotations): SexpValue<T>
+
+    abstract override fun withoutAnnotations(): SexpValue<T>
+}
+
+public abstract class StructValue<T : PartiQLValue> : PartiQLValue, Collection<Pair<String, T>> {
+
+    public abstract val fields: List<Pair<String, T>>
+
+    override val type: PartiQLValueType = PartiQLValueType.STRUCT
+
+    abstract override fun copy(annotations: Annotations): StructValue<T>
+
+    abstract override fun withAnnotations(annotations: Annotations): StructValue<T>
+
+    abstract override fun withoutAnnotations(): StructValue<T>
+}
+
+/**
+ * Any view over a PartiQLValue
+ */
+public abstract class AnyValue : PartiQLValue {
+
+    public abstract val value: PartiQLValue
+
+    override val type: PartiQLValueType
+        get() = value.type
+
+    abstract override fun copy(annotations: Annotations): AnyValue
+
+    abstract override fun withAnnotations(annotations: Annotations): AnyValue
+
+    abstract override fun withoutAnnotations(): AnyValue
+}
```

Comment:
Would be interest to know the use case of AnyValue.

Reply:
How would you model an `equals` function?